proxy/proxy.go

```diff
@@ -462,177 +346,212 @@ func (p *Proxy) lookupRoute(r *http.Request) (rt *routing.Route, params map[stri
 	return p.routing.Route(r)
 }
 
-// send a premature error response
-func sendError(w http.ResponseWriter, error string, code int) {
-	http.Error(w, error, code)
+func sendHTTPError(w http.ResponseWriter, code int) {
 	addBranding(w.Header())
+	http.Error(w, http.StatusText(code), code)
 }
 
-// http.Handler implementation
-func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	startServe := time.Now()
+// send a premature error response
+func (p *Proxy) sendError(c *context, code int) {
+	sendHTTPError(c.responseWriter, code)
 
-	start := startServe
-	rt, params := p.lookupRoute(r)
-	if rt == nil {
-		if p.flags.Debug() {
-			dbgResponse(w, &debugInfo{
-				incoming: r,
-				response: &http.Response{StatusCode: http.StatusNotFound}})
-			return
-		}
+	id := unknownRouteId
+	if c.route != nil {
+		id = c.route.Id
+	}
 
-		p.metrics.IncRoutingFailures()
-		sendError(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
-		p.metrics.MeasureServe(unknownRouteId, r.Host, r.Method, http.StatusNotFound, startServe)
-		log.Debugf("Could not find a route for %v", r.URL)
+	p.metrics.MeasureServe(
+		id,
+		c.request.Host,
+		c.request.Method,
+		code,
+		c.startServe,
+	)
+}
+
+func (p *Proxy) makeUpgradeRequest(ctx *context, route *routing.Route, req *http.Request) {
```

Comment:
no.. SPDY, web sockets. One known use case was the gcp client. Relevant: https://github.com/zalando/skipper/issues/159 and https://github.com/zalando/skipper/pull/167

I think we should decide about this feature independently from this PR.